packages/server-boilerplate/src/models/types.ts

```diff
@@ -0,0 +1,43 @@
+/**
+ * Tupaia
+ * Copyright (c) 2017 - 2021 Beyond Essential Systems Pty Ltd
+ */
+
+import { DatabaseModel, DatabaseType } from '@tupaia/database';
+
+// eslint-disable-next-line @typescript-eslint/no-explicit-any
+type UnionToIntersection<U> = (U extends any ? (k: U) => void : never) extends (k: infer I) => void
+  ? I
+  : never;
+
+// Extracts keys that have object-like values from type T
+type ObjectLikeKeys<T> = {
+  [K in keyof T]: T[K] extends Record<string, unknown> ? K : never;
+}[keyof T];
+
+// Flattens nested object to shallow object with keys joined by J
+// eg. Flatten<{ cat: { cute: true } }, '_is_'> => { cat_is_cute: true }
+type Flatten<
+  // eslint-disable-next-line @typescript-eslint/no-explicit-any
+  T extends Record<string, Record<string, any>>,
+  J extends string = '.',
+  K extends keyof T & string = keyof T & string
+> = UnionToIntersection<
+  {
+    [V in K]: { [field in keyof T[V] & string as `${V}${J}${field}`]: T[V][field] };
+  }[K]
+>;
+
+export type PartialOrArray<T> = {
+  [field in keyof T]?: T[field] extends Record<string, unknown>
+    ? PartialOrArray<T[field]>
+    : T[field] | T[field][];
+};
```

Comment:
Really we ought to have some kind of `ts-utils` package for all the utility types we've been fleshing out...

Inside server-boilerplate is okay for now I guess since all of our ts packages depend on it...